crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -232,44 +241,94 @@ impl CumulocityConverter {
             mqtt_publisher,
             service_type,
             c8y_endpoint,
+            entity_store,
         })
     }
 
+    /// Performs auto-registration process for an entity under a given
+    /// identifier.
+    ///
+    /// If an entity is a service, its device is also auto-registered if it's
+    /// not already registered.
+    ///
+    /// It returns MQTT register messages for the given entities to be published
+    /// by the mapper, so other components can also be aware of a new device
+    /// being registered.
+    fn auto_register_entity(&mut self, entity_id: &str) -> Vec<Message> {
+        let mut register_messages = vec![];
+        let (device_id, service_id) = match entity_id.split('/').collect::<Vec<&str>>()[..] {
+            ["device", device_id, "service", service_id, ..] => (device_id, Some(service_id)),
+            ["device", device_id, "", ""] => (device_id, None),
+            _ => return register_messages,
+        };
+
+        // register device if not registered
+        let device_topic = format!("{MQTT_ROOT}/device/{device_id}//");
+        if self.entity_store.get(&device_topic).is_none() {
+            let device_type = if device_id == "main" {
+                "device"
+            } else {
+                "child-device"
+            };
+            let device_name = if device_id == "main" {
+                self.entity_store.main_device_name()
+            } else {
+                device_id
+            };
+            let device_register_payload =
+                format!("{{ \"@type\":\"{device_type}\", \"@id\":\"{device_name}\"}}");
+            let device_register_message =
+                Message::new(&Topic::new(&device_topic).unwrap(), device_register_payload)
+                    .with_retain();
+            register_messages.push(device_register_message.clone());
+            self.entity_store
+                .update(device_register_message.try_into().unwrap())
+                .unwrap();
+        }
+
+        // register service itself
+        if let Some(service_id) = service_id {
+            let service_topic = format!("{MQTT_ROOT}/device/{device_id}/service/{service_id}");
+            let service_register_payload = r#"{"@type": "service", "type": "systemd"}"#.to_string();
+            let service_register_message = Message::new(
+                &Topic::new(&service_topic).unwrap(),
+                service_register_payload,
+            )
+            .with_retain();
+            register_messages.push(service_register_message.clone());
+            self.entity_store
+                .update(service_register_message.try_into().unwrap())
+                .unwrap();
+        }
+
+        register_messages
+    }
+
     fn try_convert_measurement(
         &mut self,
         input: &Message,
     ) -> Result<Vec<Message>, ConversionError> {
```

Comment:
My goal is to reach a point where this method is no more processing a raw MQTT message, but a Measurement message that has been fully decoded by the `tedge-api` and gives:

- the topic identifier of the source
- the measurement type
- the measurement value